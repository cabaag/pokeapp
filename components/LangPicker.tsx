import { Icon, Picker } from 'native-base';
import React, { useCallback, useState } from 'react';

const LangPicker: React.FC = () => {
  const [language, setLanguage] = useState('en');

  const handleChangeLang = useCallback((lang) => {
    console.log(lang);
  }, [])

  return (
    <Picker
      iosHeader="Select your language"
      iosIcon={(
        <Icon
          name="arrow-dropdown-circle"
          style={{ color: "#007aff", fontSize: 25 }}
        />
      )}
      mode="dropdown"
      onValueChange={handleChangeLang}
      selectedValue={language}
      style={{ width: undefined }}
    >
      <Picker.Item label="Wallet" value="key0" />
      <Picker.Item label="ATM Card" value="key1" />
      <Picker.Item label="Debit Card" value="key2" />
      <Picker.Item label="Credit Card" value="key3" />
      <Picker.Item label="Net Banking" value="key4" />
    </Picker>
  );
}

export default LangPicker;