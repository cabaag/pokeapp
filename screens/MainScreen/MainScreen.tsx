import { useNavigation } from '@react-navigation/native';
import Axios from 'axios';
import { Body, Button, Container, Content, Header, Icon, Input, Item, Left, Right, Spinner, Text, Switch, Picker } from 'native-base';
import React, { useCallback, useEffect, useState } from 'react';
import { Dimensions, Image, StyleSheet, View } from 'react-native';
import { TouchableOpacity } from 'react-native-gesture-handler';
import PokeCard from '../../components/PokeCard';
import { Pokemon, PokemonListResponse } from '../../types/Pokemon';
import I18n from '../../utils/i18n';


const styles = StyleSheet.create({
  logo: {
    height: 60,
    width: 140,
    flex: 1,
    alignItems: 'center',
  },
  grid: {
    flex: 1,
    flexDirection: 'row',
    flexWrap: 'wrap',
    justifyContent: 'center',
    width: Dimensions.get('window').width
  },
  col: {
    width: (Dimensions.get('window').width - 16) / 3,
    flex: 1
  },
  paginator: {
    flexDirection: "row",
    flex: 1,
    left: 0,
    right: 0,
    justifyContent: 'space-between',
  }
})

export default function MainScreen(): React.ReactElement {
  const navigation = useNavigation();
  const limit = 50;
  const [currentPage, setCurrentPage] = useState(0);
  const [count, setCount] = useState(0);
  const [pokemons, setPokemons] = useState<Pokemon[]>([]);
  const [hasNext, setHasNext] = useState(false);
  const [hasPrevious, setHasPrevious] = useState(false);
  const [loading, setLoading] = useState(false);
  const [search, setSearch] = useState('');
  const [language, setLanguage] = useState('en');

  const fetchList = useCallback((page: number) => {
    setCurrentPage(page);
    setLoading(true);
    setPokemons([]);
    Axios.get<PokemonListResponse>(`https://pokeapi.co/api/v2/pokemon?limit=${limit}&offset=${page * limit}`)
      .then(({ data }) => {

        setCount(data.count);
        setHasNext(data.next !== null)
        setHasPrevious(data.previous !== null)
        setPokemons(data.results as Pokemon[])
        setLoading(false)

        const queries = data.results.map((p, index) => {
          return Axios.get<Pokemon>(`https://pokeapi.co/api/v2/pokemon/${p.name}`)
            .then(({ data: pokemonData }) => {
              pokemons[index] = pokemonData as Pokemon;
              setPokemons([...pokemons])
            })
        })
        Promise.all(queries).finally(() => {
          setLoading(false)
        });
      })
  }, [])

  useEffect(() => {
    fetchList(1);
  }, []);


  const handleChangePage = useCallback((page) => {
    fetchList(page);
  }, [])

  const handlePressPokemon = useCallback((pokemon: Pokemon) => {
    navigation.navigate('PokemonDetailsScreen', { pokemon });
  }, [])

  const handleSearch = useCallback((newSearch: string) => {
    setSearch(newSearch)
  }, [])

  const handleChangeLang = useCallback((lang) => {
    I18n.locale = lang;
    setLanguage(lang);
  }, [])

  return (
    <Container>
      <Header>
        <Left />
        <Body style={{ width: '100%', flex: 1, alignItems: 'center', justifyContent: 'center' }}>
          <Image source={require('../../assets/images/pokemon.png')} style={styles.logo} />
        </Body>
        <Right>
          <Picker
            iosHeader="Select your language"
            iosIcon={(
              <Icon
                name="arrow-dropdown-circle"
                style={{ color: "#007aff", fontSize: 25 }}
              />
            )}
            mode="dropdown"
            onValueChange={handleChangeLang}
            selectedValue={language}
            style={{ color: "#ffffff", fontSize: 25, width: 50 }}
          >
            <Picker.Item label="English" value="en" />
            <Picker.Item label="Spanish" value="es" />
          </Picker>
        </Right>
      </Header>
      <Header rounded searchBar>
        <Item>
          <Icon name="ios-search" />
          <Input
            onChangeText={text => handleSearch(text)}
            placeholder={I18n.t('search')}
          />
        </Item>
        <Button transparent>
          <Text>Search</Text>
        </Button>
      </Header>
      <Content>
        <View style={styles.grid}>
          {
            loading && pokemons.length ?
              <Spinner /> :
              pokemons
                .filter(p => p?.name.toLowerCase().includes(search.toLowerCase()))
                .map(pokemon =>
                  (
                    <TouchableOpacity key={pokemon?.name} onPress={() => handlePressPokemon(pokemon)} style={styles.col}>
                      <PokeCard pokemon={pokemon} />
                    </TouchableOpacity>
                  )
                )
          }
        </View>
        <View style={styles.paginator}>
          <Button disabled={!hasPrevious || loading} onPress={() => handleChangePage(currentPage - 1)}>
            <Text>
              {I18n.t('previous')}
            </Text>
          </Button>
          <Button disabled={!hasNext || loading} onPress={() => handleChangePage(currentPage + 1)}>
            <Text>
              {I18n.t('next')}
            </Text>
          </Button>
        </View>
      </Content>
    </Container>
  );
}
