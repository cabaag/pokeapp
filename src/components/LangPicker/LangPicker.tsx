import { Icon, Picker } from 'native-base';
import React, { useCallback, useState, memo } from 'react';
import { useTranslation } from 'react-i18next';

const LangPicker: React.FC = () => {
  const { t, i18n } = useTranslation();
  const [language, setLanguage] = useState(i18n.language);

  const handleChangeLang = useCallback((lang) => {
    i18n.changeLanguage(lang);
    setLanguage(lang);
  }, [])

  return (
    <Picker
      iosHeader={t("select_lang")}
      iosIcon={(
        <Icon
          name="arrow-dropdown-circle"
        />
      )}
      mode="dropdown"
      onValueChange={handleChangeLang}
      placeholder={t("select_lang")}
      selectedValue={language}
      

    >
      <Picker.Item label={t("english")} value="en" />
      <Picker.Item label={t("spanish")} value="es" />
    </Picker>
  );
}

export default memo(LangPicker);