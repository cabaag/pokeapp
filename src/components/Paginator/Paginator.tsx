import { Button, Icon, Text } from 'native-base';
import React, { memo, useCallback, useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { StyleSheet, View } from 'react-native';

const styles = StyleSheet.create({
  paginator: {
    flexDirection: "row",
    flex: 1,
    left: 0,
    right: 0,
    justifyContent: 'space-between',
    padding: 8
  }
})

export interface PaginatorProps {
  count: number;
  loading: boolean;
  onChangePage: (page: number) => void;
  interval?: number;
}

const Paginator: React.FC<PaginatorProps> = ({
  count,
  loading,
  onChangePage,
  interval = 5,
}: PaginatorProps) => {

  const [currentPage, setCurrentPage] = useState<number>(0);
  const [lastPage, setLastPage] = useState(0);
  const [offset] = useState(Math.floor(interval / 2));

  const handleChangePage = useCallback((page: number) => {
    setCurrentPage(page);
    onChangePage(page);
  }, [onChangePage]);

  useEffect(() => {
    setLastPage(Math.floor(count / 50))
  }, [count])

  const renderButtons = useCallback(() => {
    let minPage: number = Math.min(...[
      currentPage - offset, Math.abs(lastPage - interval)
    ])
    minPage = minPage < 0 ? 0 : minPage;
    let maxPage: number = Math.min(...[currentPage + interval - offset, lastPage])
    maxPage = maxPage < 5 ? 5 : maxPage;
    const buttons = [];

    // eslint-disable-next-line no-plusplus
    for (let i = minPage; i < maxPage; i++) {
      buttons.push(
        <Button bordered={currentPage !== i} disabled={currentPage === i || loading} key={i} onPress={() => handleChangePage(i)} test-id={`page-button-${i}`}>
          <Text>
            {i + 1}
          </Text>
        </Button>
      )
    }
    return buttons;
  }, [lastPage, currentPage, loading])

  const prevDisabled = currentPage === 0 || loading;
  const nextDisabled = currentPage === lastPage - 1 || loading;
  return (
    <View style={styles.paginator}>
      <Button bordered disabled={prevDisabled} onPress={() => handleChangePage(0)} test-id="first">
        <Icon name="previous" type="Foundation" />
      </Button>
      { renderButtons()}
      <Button bordered disabled={nextDisabled} onPress={() => handleChangePage(lastPage - 1)} test-id="last">
        <Icon name="next" type="Foundation" />
      </Button>
    </View>
  )
}

export default memo(Paginator);