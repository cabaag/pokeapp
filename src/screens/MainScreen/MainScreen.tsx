import { useNavigation } from '@react-navigation/native';
import Axios from 'axios';
import { Button, Container, Content, Header, Icon, Input, Item, Row, Spinner, Text } from 'native-base';
import React, { useCallback, useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Dimensions, StyleSheet, View } from 'react-native';
import { TouchableOpacity } from 'react-native-gesture-handler';
import Paginator from '../../components/Paginator/Paginator';
import PokeCard from '../../components/PokeCard/PokeCard';
import { Pokemon, PokemonListResponse } from '../../types/Pokemon';

const styles = StyleSheet.create({
  grid: {
    flex: 1,
    flexDirection: 'row',
    flexWrap: 'wrap',
    justifyContent: 'flex-start',
    alignContent: 'stretch',
    alignItems: 'stretch',
    width: Dimensions.get('window').width,
    paddingLeft: 8
  },
  col: {
    width: (Dimensions.get('window').width - 16) / 3,
    flex: 1,
    paddingBottom: 8,
  },
})

export default function MainScreen(): React.ReactElement {
  const { t } = useTranslation();
  const navigation = useNavigation();
  const limit = 50;
  const [count, setCount] = useState(0);
  const [pokemons, setPokemons] = useState<Pokemon[]>([]);
  const [loading, setLoading] = useState(false);
  const [search, setSearch] = useState('');

  const fetchList = useCallback((page: number) => {
    setPokemons([]);
    setLoading(true);

    Axios.get<PokemonListResponse>(`https://pokeapi.co/api/v2/pokemon?limit=${limit}&offset=${page * limit}`)
      .then(({ data }) => {

        setCount(data.count);
        setPokemons(data.results as Pokemon[])
        setLoading(false)

        const queries = data.results.map((p, index) => {
          return Axios.get<Pokemon>(`https://pokeapi.co/api/v2/pokemon/${p.name}`)
            .then(({ data: pokemonData }) => {
              pokemons[index] = pokemonData as Pokemon;
              setPokemons([...pokemons])
            })
        })
        Promise.all(queries)
          .catch(e => console.log(e))
          .finally(() => {
            setLoading(false)
          });
      })
  }, [])

  useEffect(() => {
    fetchList(0);
  }, []);


  const handleChangePage = useCallback((page) => {
    fetchList(page);
  }, [])

  const handlePressPokemon = useCallback((pokemon: Pokemon) => {
    navigation.navigate('PokemonDetailsScreen', { pokemon });
  }, [])

  const handleSearch = useCallback((newSearch: string) => {
    setSearch(newSearch)
  }, [])

  return (
    <Container>
      <Header searchBar>
        <Item>
          <Icon name="ios-search" />
          <Input
            onChangeText={handleSearch}
            placeholder={t('search')}
          />
        </Item>
        <Button transparent>
          <Text>{t('search')}</Text>
        </Button>
      </Header>
      {/* <Header>
      </Header> */}
      <Content>
        <Paginator count={count} loading={loading} onChangePage={handleChangePage} />
        <View style={styles.grid}>
          {
            loading && !pokemons.length ? (
              <Row style={{ justifyContent: 'center' }}>
                <Spinner size={80} />
              </Row>
            ) :
              pokemons
                .filter(p => p?.name.toLowerCase().includes(search.toLowerCase()))
                .map(pokemon =>
                  (
                    <TouchableOpacity key={pokemon?.name} onPress={() => handlePressPokemon(pokemon)} style={styles.col}>
                      <PokeCard pokemon={pokemon} />
                    </TouchableOpacity>
                  )
                )
          }
        </View>
      </Content>
    </Container>
  );
}
