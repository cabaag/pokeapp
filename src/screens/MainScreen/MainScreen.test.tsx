import { shallow, ShallowWrapper } from 'enzyme';
import { Input } from 'native-base';
import React from 'react';
import LangPicker from '../../components/LangPicker/LangPicker';
import Paginator from '../../components/Paginator/Paginator';
import MainScreen from './MainScreen';


jest.mock('@react-navigation/native');
jest.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (key: string) => key,
  }),
}));

function setup(): ShallowWrapper {
  return shallow(<MainScreen />)
}

describe.only('<MainScreen />', () => {
  let wrapper: ShallowWrapper;


  beforeEach(() => {
    wrapper = setup();
  })

  it('matchs snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('handle search', () => {
    const input = wrapper.find(Input);
    input.simulate('changeText', 'test');
  })

  it('contains LangPicker', () => {
    expect(wrapper.find(LangPicker)).toHaveLength(1);
  })

  it('contains Paginator', () => {
    expect(wrapper.find(Paginator)).toHaveLength(1);
  })

});