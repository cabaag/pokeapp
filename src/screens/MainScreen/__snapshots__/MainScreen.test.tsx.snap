// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainScreen /> matchs snapshot 1`] = `
<Styled(Container)>
  <Styled(Header)
    searchBar={true}
  >
    <Styled(Item)>
      <Styled(Icon)
        name="ios-search"
      />
      <Styled(Input)
        onChangeText={[Function]}
        placeholder="search"
      />
    </Styled(Item)>
    <Styled(Button)
      transparent={true}
    >
      <Styled(Text)>
        search
      </Styled(Text)>
    </Styled(Button)>
  </Styled(Header)>
  <Styled(Content)>
    <Memo(Paginator)
      count={0}
      loading={false}
      onChangePage={[Function]}
    />
    <View
      style={
        Object {
          "alignContent": "stretch",
          "alignItems": "stretch",
          "flex": 1,
          "flexDirection": "row",
          "flexWrap": "wrap",
          "justifyContent": "flex-start",
          "paddingLeft": 8,
          "width": 750,
        }
      }
    />
  </Styled(Content)>
</Styled(Container)>
`;
